README.md
# podcasts 
[![GoDoc](https://img.shields.io/badge/go-documentation-blue.svg?style=flat-square)](https://godoc.org/github.com/jbub/podcasts) 
[![Build Status](https://cloud.drone.io/api/badges/jbub/podcasts/status.svg)](https://cloud.drone.io/jbub/podcasts)
[![Go Report Card](https://goreportcard.com/badge/github.com/jbub/podcasts)](https://goreportcard.com/report/github.com/jbub/podcasts)

Podcast generator written in Go.

## Install

```bash
go get github.com/jbub/podcasts
```

## Docs

https://godoc.org/github.com/jbub/podcasts

## Example usage

```go
package main

import (
	"log"
	"os"
	"time"

	"github.com/jbub/podcasts"
)

func main() {
	// initialize the podcast
	p := &podcasts.Podcast{
		Title:       "My podcast",
		Description: "This is my very simple podcast.",
		Language:    "EN",
		Link:        "http://www.example-podcast.com/my-podcast",
		Copyright:   "2015 My podcast copyright",
	}

	// add first podcast item
	p.AddItem(&podcasts.Item{
		Title:    "Episode 1",
		GUID:     "http://www.example-podcast.com/my-podcast/1/episode-one",
		PubDate:  podcasts.NewPubDate(time.Now()),
		Duration: podcasts.NewDuration(time.Second * 230),
		Enclosure: &podcasts.Enclosure{
			URL:    "http://www.example-podcast.com/my-podcast/1/episode.mp3",
			Length: "12312",
			Type:   "MP3",
		},
	})

	// add second podcast item
	p.AddItem(&podcasts.Item{
		Title:   "Episode 2",
		GUID:    "http://www.example-podcast.com/my-podcast/2/episode-two",
		PubDate: podcasts.NewPubDate(time.Now()),
		Duration: podcasts.NewDuration(time.Second * 320),
		Enclosure: &podcasts.Enclosure{
			URL:    "http://www.example-podcast.com/my-podcast/2/episode.mp3",
			Length: "46732",
			Type:   "MP3",
		},
	})

	// get podcast feed, you can pass options to customize it
	feed, err := p.Feed(
		podcasts.Author("Author Name"),
		podcasts.Block,
		podcasts.Explicit,
		podcasts.Complete,
		podcasts.NewFeedURL("http://www.example-podcast.com/new-feed-url"),
		podcasts.Subtitle("This is my very simple podcast subtitle."),
		podcasts.Summary("This is my very simple podcast summary."),
		podcasts.Owner("Podcast Owner", "owner@example-podcast.com"),
		podcasts.Image("http://www.example-podcast.com/my-podcast.jpg"),
	)

	// handle error
	if err != nil {
		log.Fatal(err)
	}

	// finally write the xml to any io.Writer
	feed.Write(os.Stdout)
}
```

Which gives us this XML output:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd" version="2.0">
  <channel>
    <title>My podcast</title>
    <link>http://www.example-podcast.com/my-podcast</link>
    <copyright>2015 My podcast copyright</copyright>
    <language>EN</language>
    <description>This is my very simple podcast.</description>
    <itunes:author>Author Name</itunes:author>
    <itunes:block>yes</itunes:block>
    <itunes:explicit>yes</itunes:explicit>
    <itunes:complete>yes</itunes:complete>
    <itunes:new-feed-url>http://www.example-podcast.com/new-feed-url</itunes:new-feed-url>
    <itunes:subtitle>This is my very simple podcast subtitle.</itunes:subtitle>
    <itunes:summary><![CDATA[This is my very simple podcast summary.]]></itunes:summary>
    <itunes:owner>
      <itunes:name>Podcast Owner</itunes:name>
      <itunes:email>owner@example-podcast.com</itunes:email>
    </itunes:owner>
    <itunes:image href="http://www.example-podcast.com/my-podcast.jpg"></itunes:image>
    <item>
      <title>Episode 1</title>
      <guid>http://www.example-podcast.com/my-podcast/1/episode-one</guid>
      <pubDate>Tue, 10 Nov 2009 23:00:00 +0000</pubDate>
      <itunes:duration>3:50</itunes:duration>
      <enclosure url="http://www.example-podcast.com/my-podcast/1/episode.mp3" length="12312" type="MP3"></enclosure>
    </item>
    <item>
      <title>Episode 2</title>
      <guid>http://www.example-podcast.com/my-podcast/2/episode-two</guid>
      <pubDate>Tue, 10 Nov 2009 23:00:00 +0000</pubDate>
      <itunes:duration>5:20</itunes:duration>
      <enclosure url="http://www.example-podcast.com/my-podcast/2/episode.mp3" length="46732" type="MP3"></enclosure>
    </item>
  </channel>
</rss>
```


CHANGELOG.md
## v0.2.0

- Use Go 1.17.

## v0.1.0

- Initial release.

feed_test.go
package podcasts

import (
	"encoding/xml"
	"testing"
	"time"
)

func TestPubDateMarshalling(t *testing.T) {
	tm := time.Date(2015, time.January, 1, 0, 0, 0, 0, time.UTC)
	pubDate := NewPubDate(tm)
	want := "<PubDate>Thu, 01 Jan 2015 00:00:00 +0000</PubDate>"
	out, err := xml.Marshal(pubDate)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if got := string(out); want != got {
		t.Errorf("expected %v got %v", want, got)
	}
}

func TestDurationMarshalling(t *testing.T) {
	cases := []struct {
		dur  time.Duration
		want string
	}{
		{
			dur:  0,
			want: "<Duration>0:00</Duration>",
		},
		{
			dur:  time.Second * 6,
			want: "<Duration>0:06</Duration>",
		},
		{
			dur:  time.Second * 64,
			want: "<Duration>1:04</Duration>",
		},
		{
			dur:  time.Second * 125,
			want: "<Duration>2:05</Duration>",
		},
		{
			dur:  time.Second * 3600,
			want: "<Duration>1:00:00</Duration>",
		},
		{
			dur:  time.Second * 37000,
			want: "<Duration>10:16:40</Duration>",
		},
	}

	for _, cs := range cases {
		t.Run(cs.want, func(t *testing.T) {
			dur := NewDuration(cs.dur)
			out, err := xml.Marshal(dur)
			if err != nil {
				t.Errorf("unexpected error %v", err)
			}
			if got := string(out); cs.want != got {
				t.Errorf("expected %v got %v", cs.want, got)
			}
		})
	}
}


options_test.go
package podcasts

import (
	"testing"
)

func TestAuthor(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	author := "john"
	if err := Author(author)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if author != feed.Channel.Author {
		t.Errorf("expected %v got %v", author, feed.Channel.Author)
	}
}

func TestBlock(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	if err := Block(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if ValueYes != feed.Channel.Block {
		t.Errorf("expected %v got %v", ValueYes, feed.Channel.Block)
	}
}

func TestExplicit(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	if err := Explicit(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if ValueYes != feed.Channel.Explicit {
		t.Errorf("expected %v got %v", ValueYes, feed.Channel.Explicit)
	}
}

func TestComplete(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	if err := Complete(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if ValueYes != feed.Channel.Complete {
		t.Errorf("expected %v got %v", ValueYes, feed.Channel.Complete)
	}
}

func TestNewFeedURL(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	newURL := "http://example.com/test"
	if err := NewFeedURL(newURL)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if newURL != feed.Channel.NewFeedURL {
		t.Errorf("expected %v got %v", ValueYes, feed.Channel.NewFeedURL)
	}
}

func TestNewFeedURLInvalid(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	newURL := "invalid url"
	if err := NewFeedURL(newURL)(feed); err == nil {
		t.Error("expected error")
	}
}

func TestSubtitle(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	subtitle := "this is subtitle"
	if err := Subtitle(subtitle)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if subtitle != feed.Channel.Subtitle {
		t.Errorf("expected %v got %v", subtitle, feed.Channel.Subtitle)
	}
}

func TestSummary(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	summary := `this is summary. <a href="http://example.com/more">more</a>`
	if err := Summary(summary)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if summary != feed.Channel.Summary.Value {
		t.Errorf("expected %v got %v", summary, feed.Channel.Summary.Value)
	}
}

func TestOwner(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	name := "anabelle"
	email := "test@test.com"
	if err := Owner(name, email)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if name != feed.Channel.Owner.Name {
		t.Errorf("expected %v got %v", name, feed.Channel.Owner.Name)
	}
	if email != feed.Channel.Owner.Email {
		t.Errorf("expected %v got %v", email, feed.Channel.Owner.Email)
	}
}

func TestImage(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	href := "http://example.com/test/image.jpg"
	if err := Image(href)(feed); err != nil {
		t.Errorf("unexpected error %v", err)
	}
	if href != feed.Channel.Image.Href {
		t.Errorf("expected %v got %v", href, feed.Channel.Image.Href)
	}
}

func TestImageInvalid(t *testing.T) {
	feed := &Feed{
		Channel: &Channel{},
	}
	href := "invalid img url"
	if err := Image(href)(feed); err == nil {
		t.Errorf("expected error")
	}
}


podcast_test.go
package podcasts

import (
	"bytes"
	"fmt"
	"strings"
	"testing"
	"time"
)

var (
	validItems = []struct {
		title           string
		guid            string
		pubDate         time.Time
		pubDateStr      string
		enclosureURL    string
		enclosureLength string
		enclosureType   string
		duration        time.Duration
		durationStr     string
	}{
		{
			title:           "Item 1",
			guid:            "http://www.example-podcast.com/my-podcast/1/episode",
			pubDate:         time.Date(2015, time.January, 1, 0, 0, 0, 0, time.UTC),
			pubDateStr:      "Thu, 01 Jan 2015 00:00:00 +0000",
			enclosureURL:    "http://www.example-podcast.com/my-podcast/1/episode-one",
			enclosureLength: "1234",
			enclosureType:   "MP3",
		},
		{
			title:           "Item 2",
			guid:            "http://www.example-podcast.com/my-podcast/2/episode",
			pubDate:         time.Date(2015, time.January, 2, 0, 0, 0, 0, time.UTC),
			pubDateStr:      "Fri, 02 Jan 2015 00:00:00 +0000",
			enclosureURL:    "http://www.example-podcast.com/my-podcast/2/episode-two",
			enclosureLength: "56445",
			enclosureType:   "WAV",
			duration:        time.Second * 94,
			durationStr:     "1:34",
		},
	}
)

func TestContainsXmlHeader(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := `<?xml version="1.0" encoding="UTF-8"?>`
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsRssElement(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := `<rss xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd" version="2.0">`
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsChannelElement(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := `<channel>`
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
	want = `</channel>`
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsTitleElement(t *testing.T) {
	podcast := &Podcast{Title: "my title"}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<title>%v</title>", podcast.Title)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsDescriptionElement(t *testing.T) {
	podcast := &Podcast{Description: "my description"}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<description>%v</description>", podcast.Description)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsLanguageElement(t *testing.T) {
	podcast := &Podcast{Language: "en"}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<language>%v</language>", podcast.Language)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsLinkElement(t *testing.T) {
	podcast := &Podcast{Link: "https://example.com"}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<link>%v</link>", podcast.Link)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsCopyrightElement(t *testing.T) {
	podcast := &Podcast{Copyright: "MIT"}
	data, err := getPodcastXML(podcast)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<copyright>%v</copyright>", podcast.Copyright)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsBlockElement(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast, Block)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:block>%v</itunes:block>", ValueYes)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsExplicitElement(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast, Explicit)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:explicit>%v</itunes:explicit>", ValueYes)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsCompleteElement(t *testing.T) {
	podcast := &Podcast{}
	data, err := getPodcastXML(podcast, Complete)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:complete>%v</itunes:complete>", ValueYes)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsAuthorElement(t *testing.T) {
	podcast := &Podcast{}
	author := "Test Author"
	data, err := getPodcastXML(podcast, Author(author))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:author>%v</itunes:author>", author)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsNewFeedURLElement(t *testing.T) {
	podcast := &Podcast{}
	url := "http://localhost/my-test-url"
	data, err := getPodcastXML(podcast, NewFeedURL(url))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:new-feed-url>%v</itunes:new-feed-url>", url)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsSubtitleElement(t *testing.T) {
	podcast := &Podcast{}
	subtitle := "Test Subtitle"
	data, err := getPodcastXML(podcast, Subtitle(subtitle))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:subtitle>%v</itunes:subtitle>", subtitle)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsSummaryElement(t *testing.T) {
	podcast := &Podcast{}
	summary := `Test Summary with <a href="http://www.example.com/">link</a>`
	data, err := getPodcastXML(podcast, Summary(summary))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf("<itunes:summary><![CDATA[%v]]></itunes:summary>", summary)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsOwnerElement(t *testing.T) {
	podcast := &Podcast{}
	name := "Test Name"
	email := "test@name.com"
	data, err := getPodcastXML(podcast, Owner(name, email))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := "<itunes:owner>"
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
	want = fmt.Sprintf("<itunes:name>%v</itunes:name>", name)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
	want = fmt.Sprintf("<itunes:email>%v</itunes:email>", email)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
	want = "</itunes:owner>"
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsImageElement(t *testing.T) {
	podcast := &Podcast{}
	image := "http://localhost/myimage.jpg"
	data, err := getPodcastXML(podcast, Image(image))
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	want := fmt.Sprintf(`<itunes:image href="%v"></itunes:image>`, image)
	if !strings.Contains(data, want) {
		t.Errorf("expected %v to contain %v", data, want)
	}
}

func TestContainsItemElements(t *testing.T) {
	podcast := setupPodcast()
	feed, err := podcast.Feed()
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	data, err := feed.XML()
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}

	for _, item := range validItems {
		t.Run(item.title, func(t *testing.T) {
			if want := "<item>"; !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
			if want := fmt.Sprintf("<title>%v</title>", item.title); !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
			if want := fmt.Sprintf("<guid>%v</guid>", item.guid); !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
			if want := fmt.Sprintf("<pubDate>%v</pubDate>", item.pubDateStr); !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
			if want := fmt.Sprintf(`<enclosure url="%v" length="%v" type="%v"></enclosure>`, item.enclosureURL, item.enclosureLength, item.enclosureType); !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
			if item.durationStr != "" {
				if want := fmt.Sprintf("<itunes:duration>%v</itunes:duration>", item.durationStr); !strings.Contains(data, want) {
					t.Errorf("expected %v to contain %v", data, want)
				}
			}
			if want := "</item>"; !strings.Contains(data, want) {
				t.Errorf("expected %v to contain %v", data, want)
			}
		})
	}
}

func TestPodcastFeedWrite(t *testing.T) {
	podcast := &Podcast{}
	feed, err := podcast.Feed()
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
	var b bytes.Buffer
	err = feed.Write(&b)
	if err != nil {
		t.Errorf("unexpected error %v", err)
	}
}

func getPodcastXML(p *Podcast, options ...func(f *Feed) error) (string, error) {
	feed, err := p.Feed(options...)
	if err != nil {
		return "", err
	}
	return feed.XML()
}

func setupPodcast() *Podcast {
	podcast := &Podcast{}
	for _, item := range validItems {
		podcast.AddItem(&Item{
			Title:    item.title,
			GUID:     item.guid,
			PubDate:  NewPubDate(item.pubDate),
			Duration: NewDuration(item.duration),
			Enclosure: &Enclosure{
				URL:    item.enclosureURL,
				Length: item.enclosureLength,
				Type:   item.enclosureType,
			},
		})
	}
	return podcast
}


doc.go
/*
Package podcasts implements a podcast generator.

    // initialize the podcast
    p := &podcasts.Podcast{
        Title:       "My podcast",
        Description: "This is my very simple podcast.",
        Language:    "EN",
        Link:        "http://www.example-podcast.com/my-podcast",
        Copyright:   "2015 My podcast copyright",
    }

    // add first podcast item
    p.AddItem(&podcasts.Item{
        Title:   "Episode 1",
        GUID:    "http://www.example-podcast.com/my-podcast/1/episode-one",
        PubDate: podcasts.NewPubDate(time.Now()),
        Duration: podcasts.NewDuration(time.Second * 320),
        Enclosure: &podcasts.Enclosure{
            URL:    "http://www.example-podcast.com/my-podcast/1/episode.mp3",
            Length: "12312",
            Type:   "MP3",
        },
    })

    // add second podcast item
    p.AddItem(&podcasts.Item{
        Title:   "Episode 2",
        GUID:    "http://www.example-podcast.com/my-podcast/2/episode-two",
        PubDate: podcasts.NewPubDate(time.Now()),
        Duration: podcasts.NewDuration(time.Second * 210),
        Enclosure: &podcasts.Enclosure{
            URL:    "http://www.example-podcast.com/my-podcast/2/episode.mp3",
            Length: "46732",
            Type:   "MP3",
        },
    })

    // get podcast feed, you can pass options to customize it
    feed, err := p.Feed(
        podcasts.Author("Author Name"),
        podcasts.Block,
        podcasts.Explicit,
        podcasts.Complete,
        podcasts.NewFeedURL("http://www.example-podcast.com/new-feed-url"),
        podcasts.Subtitle("This is my very simple podcast subtitle."),
        podcasts.Summary("This is my very simple podcast summary."),
        podcasts.Owner("Podcast Owner", "owner@example-podcast.com"),
        podcasts.Image("http://www.example-podcast.com/my-podcast.jpg"),
    )

    // handle error
    if err != nil {
        log.Fatal(err)
    }

    // finally write the xml to any io.Writer
    feed.Write(os.Stdout)
*/
package podcasts


options.go
package podcasts

import (
	"errors"
	"net/url"
)

var (
	// ErrInvalidURL represents a error returned for invalid url.
	ErrInvalidURL = errors.New("podcasts: invalid url")

	// ErrInvalidImage represents a error returned for invalid image.
	ErrInvalidImage = errors.New("podcasts: invalid image")
)

// Values represents positive/negative value used in XML feed.
const (
	ValueYes = "yes"
	ValueNo = "no"
)


// Author sets itunes:author of given feed.
func Author(author string) func(f *Feed) error {
	return func(f *Feed) error {
		f.Channel.Author = author
		return nil
	}
}

// Block enables itunes:block of given feed.
func Block(f *Feed) error {
	f.Channel.Block = ValueYes
	return nil
}

// Explicit enables itunes:explicit of given feed.
func Explicit(f *Feed) error {
	f.Channel.Explicit = ValueYes
	return nil
}
// NotExplicit disables itunes:explicit of given feed.
func NotExplicit(f *Feed) error {
	f.Channel.Explicit = ValueNo
	return nil
}

// Complete enables itunes:complete of given feed.
func Complete(f *Feed) error {
	f.Channel.Complete = ValueYes
	return nil
}
// NewFeedURL sets itunes:new-feed-url of given feed.
func NewFeedURL(newURL string) func(f *Feed) error {
	return func(f *Feed) error {
		u, err := url.Parse(newURL)
		if err != nil {
			return err
		}
		if !u.IsAbs() {
			return ErrInvalidURL
		}
		f.Channel.NewFeedURL = newURL
		return nil
	}
}

// Subtitle sets itunes:subtitle of given feed.
func Subtitle(subtitle string) func(f *Feed) error {
	return func(f *Feed) error {
		f.Channel.Subtitle = subtitle
		return nil
	}
}

// Summary sets itunes:summary of given feed.
func Summary(summary string) func(f *Feed) error {
	return func(f *Feed) error {
		f.Channel.Summary = &ItunesSummary{summary}
		return nil
	}
}

// Owner sets itunes:owner of given feed.
func Owner(name string, email string) func(f *Feed) error {
	return func(f *Feed) error {
		f.Channel.Owner = &ItunesOwner{
			Name:  name,
			Email: email,
		}
		return nil
	}
}

// Image sets itunes:image of given feed.
func Image(href string) func(f *Feed) error {
	return func(f *Feed) error {
		u, err := url.Parse(href)
		if err != nil {
			return err
		}
		if !u.IsAbs() {
			return ErrInvalidImage
		}
		f.Channel.Image = &ItunesImage{
			Href: href,
		}
		return nil
	}
}



// Category appends itunes:category of given feed.
// Execute multiple times to add new main category and subcategories
func Category(category string, subcategories []string) func(f *Feed) error {
	return func(f *Feed) error {
        c := &ItunesCategory{
            Text: category,
        }

        for _, subcategory := range subcategories {
            c.Categories = append(c.Categories, &ItunesCategory{
                Text: subcategory,
            })
        }

		f.Channel.Categories = append(f.Channel.Categories, c)
		return nil
	}
}


feed.go
package podcasts

import (
	"bytes"
	"encoding/xml"
	"io"
	"strconv"
	"strings"
	"time"
)

const (
	rssXmlns   = "http://www.itunes.com/dtds/podcast-1.0.dtd"
	rssVersion = "2.0"
	rfc2822    = "Mon, 02 Jan 2006 15:04:05 -0700"
)

// NewPubDate returns a new PubDate.
func NewPubDate(d time.Time) *PubDate {
	return &PubDate{d}
}

// PubDate represents pubDate attribute of given podcast item.
type PubDate struct {
	time.Time
}

// MarshalXML marshalls pubdate using the rfc2822 time format.
func (p PubDate) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	if err := e.EncodeToken(start); err != nil {
		return err
	}
	if err := e.EncodeToken(xml.CharData(p.Format(rfc2822))); err != nil {
		return err
	}
	return e.EncodeToken(xml.EndElement{Name: start.Name})
}

// NewDuration returns a new Duration.
func NewDuration(d time.Duration) *Duration {
	return &Duration{d}
}

// Duration represents itunes:duration attribute of given podcast item.
type Duration struct {
	time.Duration
}

// MarshalXML marshalls duration using HH:MM:SS, H:MM:SS, MM:SS, M:SS formats.
func (d Duration) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	if err := e.EncodeToken(start); err != nil {
		return err
	}
	if err := e.EncodeToken(xml.CharData(formatDuration(d.Duration))); err != nil {
		return err
	}
	return e.EncodeToken(xml.EndElement{Name: start.Name})
}

// formatDuration formats duration in these formats: HH:MM:SS, H:MM:SS, MM:SS, M:SS.
func formatDuration(d time.Duration) string {
	total := int(d.Seconds())
	hours := total / 3600
	total %= 3600
	minutes := total / 60
	total %= 60

	var b strings.Builder
	if hours > 0 {
		b.WriteString(strconv.Itoa(hours) + ":")
	}
	if hours > 0 && minutes < 10 {
		b.WriteString("0")
	}
	b.WriteString(strconv.Itoa(minutes) + ":")
	if total < 10 {
		b.WriteString("0")
	}
	b.WriteString(strconv.Itoa(total))
	return b.String()
}

// ItunesOwner represents the itunes:owner of given channel.
type ItunesOwner struct {
	XMLName xml.Name `xml:"itunes:owner"`
	Name    string   `xml:"itunes:name"`
	Email   string   `xml:"itunes:email"`
}

// ItunesImage represents the itunes:image of given item or channel.
type ItunesImage struct {
	XMLName xml.Name `xml:"itunes:image"`
	Href    string   `xml:"href,attr"`
}

// ItunesCategory represents itunes:category of given channel.
type ItunesCategory struct {
	XMLName    xml.Name `xml:"itunes:category"`
	Text       string   `xml:"text,attr"`
	Categories []*ItunesCategory
}

// Enclosure represents audio or video file of given item.
type Enclosure struct {
	XMLName xml.Name `xml:"enclosure"`
	URL     string   `xml:"url,attr"`
	Length  string   `xml:"length,attr,omitempty"`
	Type    string   `xml:"type,attr"`
}

// ItunesSummary represents a summary of a Channel or Item that may contain
// embedded HTML such as <a href="...">...</a> links.
type ItunesSummary struct {
	Value string `xml:",cdata"`
}

// Item represents item of given channel.
type Item struct {
	XMLName         xml.Name       `xml:"item"`
	Title           string         `xml:"title"`
	GUID            string         `xml:"guid"`
	PubDate         *PubDate       `xml:"pubDate"`
	Author          string         `xml:"itunes:author,omitempty"`
	Block           string         `xml:"itunes:block,omitempty"`
	Duration        *Duration      `xml:"itunes:duration,omitempty"`
	Explicit        string         `xml:"itunes:explicit,omitempty"`
	ClosedCaptioned string         `xml:"itunes:isClosedCaptioned,omitempty"`
	Order           int            `xml:"itunes:order,omitempty"`
	Subtitle        string         `xml:"itunes:subtitle,omitempty"`
	Summary         *ItunesSummary `xml:"itunes:summary,omitempty"`
	Enclosure       *Enclosure
	Image           *ItunesImage
}

// Channel represents a RSS channel for given podcast.
type Channel struct {
	XMLName     xml.Name       `xml:"channel"`
	Title       string         `xml:"title"`
	Link        string         `xml:"link"`
	Copyright   string         `xml:"copyright"`
	Language    string         `xml:"language"`
	Description string         `xml:"description"`
	Author      string         `xml:"itunes:author,omitempty"`
	Block       string         `xml:"itunes:block,omitempty"`
	Explicit    string         `xml:"itunes:explicit,omitempty"`
	Complete    string         `xml:"itunes:complete,omitempty"`
	NewFeedURL  string         `xml:"itunes:new-feed-url,omitempty"`
	Subtitle    string         `xml:"itunes:subtitle,omitempty"`
	Summary     *ItunesSummary `xml:"itunes:summary,omitempty"`
	Owner       *ItunesOwner
	Image       *ItunesImage
	Items       []*Item
	Categories  []*ItunesCategory
}

// Feed wraps the given RSS channel.
type Feed struct {
	XMLName xml.Name `xml:"rss"`
	Xmlns   string   `xml:"xmlns:itunes,attr"`
	Version string   `xml:"version,attr"`
	Channel *Channel
}

// SetOptions sets options of given feed.
func (f *Feed) SetOptions(options ...func(f *Feed) error) error {
	for _, opt := range options {
		if err := opt(f); err != nil {
			return err
		}
	}
	return nil
}

// XML marshalls feed to XML string.
func (f *Feed) XML() (string, error) {
	var buf bytes.Buffer
	if err := f.Write(&buf); err != nil {
		return "", err
	}
	return buf.String(), nil
}

// Write writes marshalled XML to the given writer.
func (f *Feed) Write(w io.Writer) error {
	if _, err := w.Write([]byte(xml.Header)); err != nil {
		return err
	}
	enc := xml.NewEncoder(w)
	enc.Indent("", "  ")
	return enc.Encode(f)
}
